### PR TITLE
[posix-trel] support TREL URL

### PR DESCRIPTION
**Pull request by simonlingoogle**
This commit supports TREL URL (e.g. `trel://eth0`) for Posix.

Depends-On: openthread/ot-br-posix#803

### CODE REVIEW

**Comment by abtink** [comment]
This change here to use URL looks fine to me.

However, there is one point that makes me a bit uneasy for merging this. I'd like to ask that we postpone/delay merging this PR (which is related to TREL posix implementation) before we have added back some tests or at minimum some build checks in our CI for posix build with TREL enabled.

From issue https://github.com/openthread/openthread/issues/6472 and removal of `ot-ncp` (which also removed the related `toranj` test-cases that were covering TREL poisx build), I don't think we have  any CI converge for posix BUILD with TREL platform enabled. I think we need at least add back a CI build check under posix in which TREL is enabled (before we merge changes to the related code).

**Comment by abtink** [comment]
Another point on radio URL use to have in mind:
- A radio URL so far has been used as a way to specify and configure a module that acts as a radio platform for OT core, i.e., a module that implements a radio driver and provides `otPlatRadio` APIs.
- The TREL does not actually define and/or provide an `otPlatRadio` abstraction. It defines its own platform abstraction (which is way different from `otPlatRadio` definitions).  I think it is important to not think of TREL as a module that provides `otPlatRadio` APIs 
    - A side-note: I know at first it may seem reasonable to think of multi-radio as having multiple implementations of `otPlatRadio` APIs but it does not really work and does not allow the behaviors we want from each PHY. 
    - The reason for this is because `otPlatRadio` APIs are too low-level and very much tied to 15.4 PHY behaviors. The right point for handling multi-radio is at a higher abstraction layer (`Mac` layer turned out to be proper point for handling multi-radio and that's how it is implemented in our OT core implementation).

**Comment by simonlingoogle** [comment]
> From issue #6472 and removal of `ot-ncp` (which also removed the related `toranj` test-cases that were covering TREL poisx build), I don't think we have any CI converge for posix BUILD with TREL platform enabled. I think we need at least add back a CI build check under posix in which TREL is enabled (before we merge changes to the related code).

This PR is required by https://github.com/openthread/ot-br-posix/pull/803. The very purpose of this PR is to enable TREL in OT Posix and ot-br-posix (i.e. otbr-agent) in a unified way, so that we can test TREL link in Border Router tests.

**Comment by simonlingoogle** [comment]
> * The TREL does not actually define and/or provide an `otPlatRadio` abstraction.

Indeed. However, using TREL URL does not necessarily means that the OT core should treat them equally. The purpose is just to provide a unified and extensible way of specifying radio links. For example, it's easy to extend the TREL URL to include parameters for future extension (e.g. `trel::/eth0?desirable=false&use-mdns=true`).

**Comment by abtink** [comment]
> This PR is required by openthread/ot-br-posix#803. The very purpose of this PR is to enable TREL in OT Posix and ot-br-posix (i.e. otbr-agent) in a unified way, so that we can test TREL link in Border Router tests.

Having a test in `ot-br-posix` is good, but still I think we should at minimum have a build check in `opethread` CI itself as well (for posix build with TREL). It will be useful to make sure future changes/PR won't break the build

I will try to submit a PR to add such a check.

**Comment by simonlingoogle** [comment]
> I will try to submit a PR to add such a check.

Thanks! That would be very helpful.

**Comment by simonlingoogle** [comment]
Any further comments on this PR? @bukepo @abtink

**Review by bukepo — APPROVED** [review]
LGTM:+1:

**Comment by abtink** [comment]
> Any further comments on this PR? @bukepo @abtink

We need to wait for https://github.com/openthread/openthread/pull/6518 to be merged, then please rebase this PR on it (so we have some CI build checks covering posix TREL and chnaged in this PR).

**Comment by simonlingoogle** [comment]
**Merge order tip:** @jwhui 
1. Merge this PR first
1. Merge openthread/ot-br-posix#803 with updated openthread repo

**Review by abtink — APPROVED** [review]
LGTM. Thanks 👍

File: src/posix/platform/system.cpp [hunk 1]
```diff
@@ -66,10 +66,51 @@ static void processStateChange(otChangedFlags aFlags, void *aContext)
...
 #endif
 
+static const char *get802154RadioUrl(otPlatformConfig *aPlatformConfig)
+{
+    const char *radioUrl = nullptr;
+
+    for (uint8_t i = 0; i < aPlatformConfig->mRadioUrlNum; i++)
+    {
+        ot::Posix::RadioUrl url(aPlatformConfig->mRadioUrls[i]);
+        if (strcmp(url.GetProtocol(), "trel") == 0)
```
Comment (jwhui) [inline src/posix/platform/system.cpp:76]: nit: add empty line after declaration
```suggestion
        ot::Posix::RadioUrl url(aPlatformConfig->mRadioUrls[i]);

        if (strcmp(url.GetProtocol(), "trel") == 0)
```

File: src/posix/platform/system.cpp [hunk 2]
```diff
@@ -66,10 +66,51 @@ static void processStateChange(otChangedFlags aFlags, void *aContext)
...
+        {
+            continue;
+        }
+        else
+        {
+            radioUrl = aPlatformConfig->mRadioUrls[i];
+            break;
+        }
```
Comment (jwhui) [inline src/posix/platform/system.cpp:84]: nit: remove unnecessary indent
```suggestion

        radioUrl = aPlatformConfig->mRadioUrls[i];
        break;
```

File: src/posix/platform/system.cpp [hunk 3]
```diff
@@ -66,10 +66,51 @@ static void processStateChange(otChangedFlags aFlags, void *aContext)
...
+
+#if OPENTHREAD_CONFIG_RADIO_LINK_TREL_ENABLE
+static const char *getTrelRadioUrl(otPlatformConfig *aPlatformConfig)
+{
+    const char *radioUrl = nullptr;
+
+    for (uint8_t i = 0; i < aPlatformConfig->mRadioUrlNum; i++)
+    {
+        ot::Posix::RadioUrl url(aPlatformConfig->mRadioUrls[i]);
+        if (strcmp(url.GetProtocol(), "trel") == 0)
```
Comment (jwhui) [inline src/posix/platform/system.cpp:99]: nit: add empty line after declaration
```suggestion
        ot::Posix::RadioUrl url(aPlatformConfig->mRadioUrls[i]);

        if (strcmp(url.GetProtocol(), "trel") == 0)
```

**Review by jwhui — APPROVED** [review]

**Pull request merged**